drivers: video: stm32u5x dcmi: enable dcmi support for stm32u5x family

cornway opened the pull request:
Adding missing logic to handle stm32u5x dcmi interface

avolmat-st commented:
Hi @cornway and thank you for this patch.
Could you explain the reason why specifically forcing the DCMI mode into snapshot here ?
FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet.
I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after.
Thank you again.

cornway commented:
> Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.



> Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.

I need more time to configure GPDMA to run in circular mode, otherwise shapshot is just a short cut
Overall it runs continuously, since I restart dma once upon frame completion
Thank you 
Please let me know if this needs to be added in the same PR

avolmat-st commented:
> > Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.
> 
> > Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.
> 
> I need more time to configure GPDMA to run in circular mode, otherwise shapshot is just a short cut Overall it runs continuously, since I restart dma once upon frame completion Thank you Please let me know if this needs to be added in the same PR

Thanks for the reply. In that case, it is better to have it run in continuous mode now to avoid introducing such U5 specific behavior code in the driver.

cornway commented:
> Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.



> Hi @cornway and thank you for this patch. Could you explain the reason why specifically forcing the DCMI mode into snapshot here ? FYI, there is currently an ongoing activity about making it possible to run in snapshot mode (cf PR #93797) so I think this might fit for your needs, even it isn't finalized yet. I haven't yet looked at the other part DT / DMA configuration in detail hence will come back to you after. Thank you again.

I need more time to configure GPDMA to run in circular mode, otherwise shapshot is just a short cut
Overall it runs continuously, since I restart dma once upon frame completion
Thank you 
Please let me know if this needs to be added in the same PR

> Hello @cornway, and thank you very much for your first pull request to the Zephyr project! Our Continuous Integration pipeline will execute a series of checks on your Pull Request commit messages and code, and you are expected to address any failures by updating the PR. Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request. If you haven't already, please make sure to review the project's [Contributor Expectations](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html) and update (by amending and force-pushing the commits) your pull request if necessary. If you are stuck or need help please join us on [Discord](https://chat.zephyrproject.org/) and ask your question there. Additionally, you can [escalate the review](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-review-escalation) when applicable. 😊

Updated PR with removed shapshot mode (Added gpdma LL for sicrcular mode)
Although it still needs cleanup it is ready for review I think
Question: do you use any formatting tools like clang, as tutorial says this is mostly manual process
Thank you

cornway committed:
drivers: video: stm32u5x dcmi: enable dcmi support for stm32u5x family

Adding missing logic to handle stm32u5x dcmi interface
Adding JPEG mode support:
affects all soc's for which dcmi has already been implemented


Signed-off-by: Roman Pustobaiev <romanpustobayev@gmail.com>

cornway commented:
@JarmouniA @avolmat-st 
Kindly ask you to review this patch
Thank you

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -109,6 +153,44 @@ void HAL_DMA_ErrorCallback(DMA_HandleTypeDef *hdma)
+	pNodeConfig.DstAddress = 0;
+	pNodeConfig.DataSize = 0;
+
+	ret |= HAL_DMAEx_List_BuildNode(&pNodeConfig, &Node);
+	ret |= HAL_DMAEx_List_InsertNode_Tail(&Queue, &Node);
+	ret |= HAL_DMAEx_List_SetCircularMode(&Queue);
+	ret |= HAL_DMAEx_List_LinkQ(hdma, &Queue);
+
+	return ret;
```
using `ret |= ` is very compact, but if there are different error messages coming, then `return ret;` would become inaccurate, and then code that does `do { ret = try_again(); } while (ret == -EAGAIN);` can behave differently.

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -84,6 +84,50 @@ void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
+
+resume:
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/*
+	 * Stop DMA as in JPEG mode
+	 * it may be still waiting for more data to come, so we can't restart it directly
+	 */
```
Is that really specific to `STM32U5` or can it be used with other SoCs?

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -109,6 +153,44 @@ void HAL_DMA_ErrorCallback(DMA_HandleTypeDef *hdma)
+	HAL_StatusTypeDef ret = HAL_OK;
+	DMA_NodeConfTypeDef pNodeConfig;
+
+	/* Set node configuration ################################################*/
```
Every vendor/author might have their own decoration style, how about a plain comment `/* Set node configuration */`?

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -283,9 +408,12 @@ static int video_stm32_dcmi_set_stream(const struct device *dev, bool enable,
+	uint32_t buffer_size = data->fmt.pitch * data->fmt.height;
 
-	if (buffer_size > vbuf->size) {
+	/* Size of a framebuffer is unknown at the moment, assign maximum */
+	if (data->fmt.pixelformat == VIDEO_PIX_FMT_JPEG) {
+		buffer_size = vbuf->size;
+	} else if (buffer_size > vbuf->size) {
```
I would be tempted to split the JPEG support in a different commit.

- The commit short subject does not talk about JPEG so harder to look at when it was introduced.
- Sometimes commits are reverted upstream to fix CI or locally troubleshoot. Then either JPEG or STM32U5 would be reverted but not both.

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -174,6 +281,20 @@ static int stm32_dma_init(const struct device *dev)
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	if (HAL_DMAEx_List_Init(&hdma) != HAL_OK) {
+		LOG_ERR("HAL_DMAEx_List_Init Failed");
+		return -EINVAL;
+	}
+	if (stm32_dma_list_init(&hdma) != HAL_OK) {
+		LOG_ERR("stm32_dma_list_init Failed");
+		return -EINVAL;
+	}
+	if (HAL_DMA_ConfigChannelAttributes(&hdma, DMA_CHANNEL_NPRIV) != HAL_OK) {
+		LOG_ERR("HAL_DMA_ConfigChannelAttributes Failed");
+		return -EINVAL;
+	}
```
Just noticing the different style through the source, in case you were interested in smoothing it up:
- above uses `err = HAL_...(); if (err != HAL_OK) {}`
- here it uses `if (HAL_...() != HAL_OK) {}`

josuah reviewed (commented):
Thanks for adding support for more hardware! Some feedback to try to take this PR further up.

erwango commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -164,6 +269,8 @@ static int stm32_dma_init(const struct device *dev)
 						config->dma.channel);
 #elif defined(CONFIG_SOC_SERIES_STM32L4X)
 	hdma.Instance = __LL_DMA_GET_CHANNEL_INSTANCE(config->dma.reg, config->dma.channel);
+#elif defined(CONFIG_SOC_SERIES_STM32U5X)
+	hdma.Instance = LL_DMA_GET_CHANNEL_INSTANCE(config->dma.reg, config->dma.channel);
```
This shouldn't be required after rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/94869

erwango commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -174,6 +281,20 @@ static int stm32_dma_init(const struct device *dev)
 		return -EIO;
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
```
I suggest checking if this DMAU5 compat shouldn't be check instead of U5 series.
